### DISCUSSION TITLE
Playing animation only once

### DISCUSSION THREAD

**Discussion by dhruvadhia1** [google/model-viewer#2226] [Q&A]
### Description

Playing animation only once after clicking play for example instead of constant looping. One idea I had is that if I can get total duration of animation then too I can stop animation from playing once it reaches the end. But it would also be good to have method like playOnce().

#### Live Demo

https://glitch.com/edit/#!/model-viewer

https://thecela.com/configurator/TEo1Y29KUzBuSjlocVRJbXFOPT0=/QVF0PQ==/cDJBdW90PT0=

### Version

Latest version

### Browser Affected

All
### OS

All

**Comment by dhruvadhia1** [comment]
@elalish I have a questions regarding events. Do we have something like 'timeudpate' event like audio html tag? In that way if I can get the value of currentTime in realtime then it is possible to play only once. When current time arrives at 0 again, then we can pause the animation and the current time can be reset again.

example : https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_av_event_timeupdate_el

**Comment by elalish** [comment]
No, for now your best bet is `setTimeout()`.

**Comment by dhruvadhia1** [comment]
Here is a way i got it working, if others are looking for solutions. 

```
 // non looping animation
	  const modelViewer = document.querySelector("model-viewer");
	  modelViewer.animationName = 'All Animations';
	  var extremelySmallIntervalTime = 0.05;
	  var j=0;
	  
	  setInterval(function(){ 
       
		  //console.log(modelViewer.currentTime);	
		  if(modelViewer.currentTime > 0 && modelViewer.currentTime < extremelySmallIntervalTime)
		  {
			   
			  if ( j > 0 )
 			  {
 				  modelViewer.pause();
 				  modelViewer.currentTime = 0;
 			  }
			   
			   j++;
			   console.log(j);
		  }
  	  }, extremelySmallIntervalTime*1000);
```

Its not perfect. but works.

**Reply by WadeMorrison** [reply]
@dhruvadhia1 Thank you so much for this! It works great minus one small issue, I have two animations (fired by two individual buttons ("Open" and "Close")) and for some reason the product won't "Open" unless I click the "Close" button first (which actually opens it) and then after that the buttons work how they're supposed to. I do have the 'All Animations' set to 'DoorOpen' and 'DoorClose' respectively. 

My animations also "snap" to the end instead of playing the full .5s track. Is this supposed to be how it works (if so it's fine, but actually having it gradually open/close would look nicer for application)

Any help is greatly appreciated!

**Comment by sercandemirtas** [comment]
@dhruvadhia1 hey man. I added this codes to js but It is not working. Animation still in loop. Any other solutions ?

**Comment by elalish** [comment]
How about something like:

```
const animationLength = 1000; // ms
const modelViewer = document.querySelector("#your-id");
modelViewer.play();
self.setTimeout(()=>{modelViewer.pause();}, animationLength);
```

**Comment by dhruvadhia1** [comment]
@elalish , this code will pause any given animation after 1 second. However, the length of the animation could vary. 

@sercandemirtas 

try this code 


```
modelViewer.play();

  var extremelySmallIntervalTime = 0.05;
  var j=0;
  const interval = setInterval(function(){
  console.log(modelViewer.currentTime);
  if(modelViewer.currentTime > 0 && modelViewer.currentTime < (extremelySmallIntervalTime - 0.01))
  { ++j;
  console.log(j);
  if ( j > 0 ){
  modelViewer.pause();
  //modelViewer.currentTime = 0;
  j = 0;
  clearInterval(interval);
  }
   
  }else if(modelViewer.currentTime == 0)
  {
  modelViewer.play();
  }
  }, extremelySmallIntervalTime*1000);
```

**Comment by sercandemirtas** [comment]
@elalish @dhruvadhia1 I tried both of your codes but animation still  in loop maybe Im doing sth wrong or there is problem with codes. Did you try this codes on your project? If you dont mind can I see project?

**Comment by dhruvadhia1** [comment]
try this link

https://thecela.com/configurator/TEo1Y29GMTBNS0EwWUdWPQ==/QVF0PQ==/cDJBdW90PT0=#

when you click on different hotspots, it will play animation either in loop or once.

**Comment by archeo56** [comment]
I use a "setTimeout" to play and stop at the end

```
<script type="module" src="https://unpkg.com/@google/model-viewer/dist/model-viewer.js"></script>
<script nomodule src="https://unpkg.com/@google/model-viewer/dist/model-viewer-legacy.js"></script> 

<div style ="height: 50px; width: 50px; border-style : outset; border-width:1px;border-radius : 10px;box-shadow :3px 3px 3px #969696 ;padding:1px 10px 0px 10px;cursor:pointer;" onclick="play_to_end()">Try it</div>

<model-viewer id="homme"  camera-controls  style="width: 400px; height: 400px;" interaction-prompt="none" src="{root_url}/uploads/images/histoire/homme/cranes/scene(20).glb" alt="">

</model-viewer>

<script>
function play_to_end (){
//play to end or rewind if animation is at end//
	const modelViewer = document.querySelector('#homme');
	modelViewer.play();
	t=modelViewer.duration;
	t=t*1000/2 ;// because duration is in second and time to play and rewind//
	setTimeout(stop, t);
}
function stop(){
const modelViewer = document.querySelector('#homme');
modelViewer.pause();
}
</script>
```

you can see the result here [http://www.evolution-biologique.org/test-anim.html](url)

**Comment by retailleaumaureen** [comment]
Hello. I try to control an animation exported in GLB with your code but at the end of the anime it reboots. I wish I had a button to play it and a button to rewind it
https://glitch.com/edit/#!/test-dimitro-d15?path=index.html%3A81%3A54
thanks a lot for you're help

**Reply by elalish** [reply]
It sounds like you may just have a  rounding  error, since any time past the end of the animation  will start again at the  beginning. You need to ensure you call pause before the time gets past the duration. We should probably consider extra  APIs here.

**Reply by retailleaumaureen** [reply]
Thank you for taking the time to help me. I am a 3d inforgraphiste and have no coding training, I learn on my own. Can you give me more indications?
thanks in advance

**Comment by uprealdigital** [comment]
Hi, 

The setTimeout solution works really well on the Webpage itself. But when going in AR mode, it doesn't work consistently.

The problem seems to be apparent when the framerate fluctuates... Is it possible that the 3D engine that holds the animations/3D models, and the web page that holds the setTimeout function are running on two different internal timers?

So if the 3D engine drops frames then the webpage keeps going then they become out of sync.

**Comment by futahei** [comment]
I'm working on implementing this feature (#3021). Please support me if you can.

**Comment by buddy90210** [comment]
It's my solution. And it's work fine.
```
$('[data-selector="change-animation"]').click(async function () {
                    let animation_name = this.dataset.name;
                    modelViewer.animationName = animation_name;
                    let updatedViewer = await getUpdatedModelViewer();
                    updatedViewer.play();
                    self.setTimeout(() => { updatedViewer.pause(); }, updatedViewer.duration*1000);
 });
function getModelViewer() {
        return document.querySelector('model-viewer.model-view');
}
async function getUpdatedModelViewer() {
        const viewer = getModelViewer();
        await viewer.updateComplete;
        return viewer;
}
```
